### ISSUE TITLE
Problems of big image files and metadata computing: multi-threading may be necessary to not block the UI

### ISSUE THREAD

**Issue by ptaillandier** [gama-platform/gama#1523]
I have a big raster file (tif) and when I copy it to one of the my workspace folder, GAMA seems to freeze (I think it try to build the meta-data, but the process is just so slow).

here a link to get my raster file: https://drive.google.com/file/d/0BxvMwvAG0APzMFRqRGV4V1RYaXM/view?usp=sharing

**Comment by AlexisDrogoul** [comment]
Probably very very slow… This process should be deported in a thread, but I'm just too lazy to do it immediately. The workaround is to turn off metadata display in the preferences (or in the "Advanced" preferences, to turn off the decorator).

**Comment by AlexisDrogoul** [comment]
I just noticed the size of the file: 572 Mb. Ahem. I understand why the process of computing meta-data chokes on this… Aren't there any way to get these information (size, dimensions, type) without reading everything ? I think it is already in place for JPG, PNG, etc, and maybe this would solve the problem if we can find one for TIFF files as well. Doesn't the library you have used provide this sort of service ?

 What I could also implement is a multi-threaded meta-data decorator, which would just request the creation of the meta-data if it is not available, in another thread, and leave immediately (displaying something like "(computing meta-data)"), setting itself a a listener of the thread to be notified when the process is over.

**Comment by AlexisDrogoul** [comment]
Actually, I tried it too, and it resulted in several exceptions:

`java.lang.OutOfMemoryError: Java heap space
	at org.eclipse.core.internal.content.LazyInputStream.loadBlock(LazyInputStream.java:105)
	at org.eclipse.core.internal.content.LazyInputStream.ensureAvailable(LazyInputStream.java:65)
	at org.eclipse.core.internal.content.LazyInputStream.read(LazyInputStream.java:139)
	at org.eclipse.swt.internal.image.LEDataInputStream.readData(LEDataInputStream.java:141)
	at org.eclipse.swt.internal.image.LEDataInputStream.read(LEDataInputStream.java:93)
	at org.eclipse.swt.internal.image.TIFFRandomFileAccess.seek(TIFFRandomFileAccess.java:46)
	at org.eclipse.swt.internal.image.TIFFFileFormat.loadFromByteStream(TIFFFileFormat.java:59)
	at org.eclipse.swt.internal.image.FileFormat.loadFromStream(FileFormat.java:55)
	at org.eclipse.swt.internal.image.FileFormat.load(FileFormat.java:86)
	at org.eclipse.swt.graphics.ImageLoader.load(ImageLoader.java:146)
	at org.eclipse.swt.graphics.ImageDataLoader.load(ImageDataLoader.java:22)
	at org.eclipse.swt.graphics.ImageData.<init>(ImageData.java:331)
	at msi.gama.gui.navigator.images.ImageDataLoader.getImageData(ImageDataLoader.java:39)
	at msi.gama.gui.navigator.images.ImageContentTypeDescriber.describe(ImageContentTypeDescriber.java:63)
	at org.eclipse.core.internal.content.ContentType.describe(ContentType.java:166)
	at org.eclipse.core.internal.content.ContentType.internalGetDescriptionFor(ContentType.java:477)
	at org.eclipse.core.internal.content.ContentTypeCatalog.getDescriptionFor(ContentTypeCatalog.java:367)
	at org.eclipse.core.internal.content.ContentTypeCatalog.getDescriptionFor(ContentTypeCatalog.java:371)
	at org.eclipse.core.internal.content.ContentTypeMatcher.getDescriptionFor(ContentTypeMatcher.java:91)
	at org.eclipse.core.internal.resources.ContentDescriptionManager.readDescription(ContentDescriptionManager.java:454)
	at org.eclipse.core.internal.resources.ContentDescriptionManager.getDescriptionFor(ContentDescriptionManager.java:364)
	at org.eclipse.core.internal.resources.File.getContentDescription(File.java:255)
	at msi.gama.gui.navigator.FileMetaDataProvider.getContentTypeId(FileMetaDataProvider.java:424)
	at msi.gama.gui.navigator.NavigatorLabelProvider.getImage(NavigatorLabelProvider.java:64)
	at org.eclipse.ui.internal.navigator.extensions.SafeDelegateCommonLabelProvider.getImage(SafeDelegateCommonLabelProvider.java:107)
	at org.eclipse.ui.internal.navigator.NavigatorContentServiceLabelProvider.findImage(NavigatorContentServiceLabelProvider.java:199)
	at org.eclipse.ui.internal.navigator.NavigatorContentServiceLabelProvider.getColumnImage(NavigatorContentServiceLabelProvider.java:107)
	at org.eclipse.ui.internal.navigator.NavigatorContentServiceLabelProvider.getImage(NavigatorContentServiceLabelProvider.java:99)
	at org.eclipse.ui.internal.navigator.NavigatorDecoratingLabelProvider$StyledLabelProviderAdapter.getImage(NavigatorDecoratingLabelProvider.java:57)
	at org.eclipse.jface.viewers.DelegatingStyledCellLabelProvider.getImage(DelegatingStyledCellLabelProvider.java:195)
	at org.eclipse.jface.viewers.DecoratingStyledCellLabelProvider.getImage(DecoratingStyledCellLabelProvider.java:173)
	at org.eclipse.jface.viewers.DelegatingStyledCellLabelProvider.update(DelegatingStyledCellLabelProvider.java:121)
GAMA Caught a workbench message : Could not create content describer for msi.gama.gui.images.type. Content type has been disabled.
GAMA Caught a workbench message : Unhandled event loop exception
org.eclipse.swt.SWTException: Invalid image (java.lang.ArrayIndexOutOfBoundsException)
org.eclipse.swt.SWTException: Invalid image (java.lang.ArrayIndexOutOfBoundsException)
null image data`


So basically, the problem is that the format of the image is not correctly read (not recognized) and that it generates an out of memory error. As a matter of fact, on MacOS X, neither Preview nor Graphic Converter can read it correctly. Only QGis succeeds — do you have any idea on the specific format of this file ?

**Comment by ptaillandier** [comment]
It is a geofiff <https://en.wikipedia.org/wiki/GeoTIFF> generated by QGIS
(using GDAL <http://www.gdal.org>).


2016-02-04 14:02 GMT+01:00 Alexis Drogoul <notifications@github.com>:

**Comment by AlexisDrogoul** [comment]
No wonder QGIS can read it, then :) I suppose, in these cases, that the best solution would be :

1. to detect these GeoTiff files (probably using their header ?)
2. to enable GDAL through GeoTools and read these files with their various adapters.

**Comment by AlexisDrogoul** [comment]
The issue is closed — although the fix does not solve the fact that GAMA cannot read this particular type of tiff image.

**Comment by AlexisDrogoul** [comment]
I have managed to, at least, be able to invoke some TIFF reader present in GeoTools, but, although it produces "something", it is a black image. The biggest problem comes from the fact that the colors are coded on 48 bits, which is something that SWT cannot deal with. So, basically, there is no way to display it, although we get the correct metadata.
